src/main/java/com/chimbori/crux/articles/PostprocessHelpers.java

```diff
@@ -203,4 +208,13 @@ private static void removeDisallowedAttributes(Element node) {
     }
   }
 
+  private static boolean isImportant(Node node) {
+      boolean result=false;
+      for(Node n=node;n!=null;n=n.parentNode())
```

Comment:
Interesting, I’m guessing you’ve run into issues where the child node was removed even though the parent node was marked important? Any thoughts about how this might affect performance, since it’s walking up the entire tree for every single node?

Reply:
> Interesting, I’m guessing you’ve run into issues where the child node was removed even though the parent node was marked important?

That's exactly right. We treat "keep this" to mean "keep this and all its children." The only way we've found to avoid discarding a node in all cases is by walking up to the root and making sure none of the ancestors are marked.

> Any thoughts about how this might affect performance, since it’s walking up the entire tree for every single node?

We haven't noticed any major slowdown in our implementation, but we haven't done any measurement either. We were comfortable with this naive implementation because it only goes back to the root on the scraped element, which is typically a much smaller (and shallower) than the whole document.

However, none of this means that it's not slower -- just that it's not noticeably slower for our use case. We can't rule out that this would be a bottleneck for someone else's use case with a different usage pattern.

Once people have marked any node as "must keep," I think it's reasonable to ask that they pay a price -- ideally small -- for the feature. However, for people not using the feature, we'd prefer there was no cost at all. What about something that looks like this:

    // Build a set of nodes for build here
    Set<Node> keepers=Collections.newSetFromMap(new IdentityHashMap<>());
    for(Element marked : topNode.select("[crux-keep=true]"))
        keepers.add(marked);

    // Our new isImportant() implementation
    private static boolean isImportant(Set<Node> keepers, Node node) {
        if(keepers.isEmpty())
            return false;
        for(Node n=node;n!=null;n=n.parentNode())
            if(keepers.contains(n))
                return true;
        return false;
    }

Here, we create a set of nodes that must be kept. We use an `IdentityHashMap` because it's very fast and doesn't rely on `Node` implementing `hashCode`, which is good because it does not. If that set is empty -- in other words, if the user didn't use the feature -- then there is essentially zero performance impact: a couple of method calls and an `if`, which the JIT will inline aggressively if it ends up being a bottleneck. If that set is not empty -- in other words, if the user did use the feature -- then the performance impact is minimized for a modest memory cost. This approach would work well as long as we agree never to clone a node, or mark any after that set is built. 

This approach does require carrying around a `Set`. To simplify things, I would probably make `PostprocessHelpers` an instance class (rather than a collection of static methods) and treat it like this:

    PostprocessHelpers postprocessor=PostprocessHelpers.train(topNode);
    postprocessor.doStep1(topNode);
    postprocessor.doStep2(topNode);
    // ...and so on.

That way, `PostprocessHelpers` can manage all the details of how it needs to initialize the postprocessing. We don't have to spread that logic out across all the users of this code.

For symmetry, I'd probably do the same to `PreprocessHelpers`, too, but that's down to style rather than substance.

There is some complexity here, but not much. Would you want to see a version with this change?

Reply:
Going through this one more time, I'm realizing you actually have to do more work. We're checking *up*, but not *down*. There's nothing stopping crux from deleting a node that *contains* a keeper, and right now we're not checking for that. I've updated the PR to check children, too. Given that cost, I think an improvement is wise. What do you think of the above idea?

Reply:
That detailed plan sounds good, thank you for looking into it some more!

If you’d rather submit that as a separate PR, that would be OK too. I will wait to issue a new release until all the related PRs are merged. Does that sound good to you?

Reply:
Yes, that makes a lot of sense. Thank you for working with me to get this feature in! 😄